### PR TITLE
[Cursor Support] Massive Monolith Commit

### PR DESCRIPTION
**Pull request by markdirish**
Create a Cursor class to allow developers to fetch n amount of results instead of fetching all
Create ODBC cursor class implementation
Create JavaScript cursor class wrapper
Rewrite entire bind/fetch/transfer-to-NAPI logic to handle multiple-rows being returned using column-wise binding
Rename 'QueryData' to 'StatementData'
Create method of passing options to queries instead of just connections

TODO:
~Write tests for all of the above~
~Figure a way to pass the parameter array to cursors (so it can be returned on the result)~
~Make sure Cursor.close works as intended~
~Test that multiple cursor works~
~Make sure that it compiles with `DEBUG` enabled~

Signed-off-by: Mark Irish <mirish@ibm.com>

### CODE REVIEW

**Comment by markdirish** [comment]
Obviously the giant monolith commit isn't ideal, but wanted to wait to push until I had something that could at least run. And it required a lot of re-plumbing to get that done.

**Comment by markdirish** [comment]
Finally figured out how to pass the parameters to the cursor! So time to clean everything up, then get this merged into the 2.4.0 branch

File: src/odbc_connection.cpp [hunk 1]
```diff
@@ -674,14 +471,156 @@ Napi::Value ODBCConnection::CreateStatement(const Napi::CallbackInfo& info) {
...
 
+typedef struct QueryOptions {
+  bool         use_cursor;
+  SQLTCHAR    *cursor_name;
+  SQLSMALLINT  cursor_name_length;
+  SQLULEN      fetch_size;
+} QueryOptions;
+
+#define QUERY_OPTION_PROPERTY_NAME_CURSOR     "cursor"
+#define QUERY_OPTION_PROPERTY_NAME_FETCH_SIZE "fetchSize"
```
Comment (kadler) [inline src/odbc_connection.cpp:482]: These could be made static class constants:

```c++
struct QueryOptions {
  // ...
  static constexpr const char* cursor_prop = "cursor";
};

QueryOptions parse_query_options() {
  if options_object.HasOwnProperty(QueryOptions::cursor_prop)) {
  }
}
```

File: src/odbc_connection.cpp [hunk 2]
```diff
@@ -674,14 +471,156 @@ Napi::Value ODBCConnection::CreateStatement(const Napi::CallbackInfo& info) {
...
+      query_options.cursor_name_length = cursor_string.length();
+    }
+    else if (cursor_value.IsBoolean())
+    {
+      query_options.use_cursor  = cursor_value.As<Napi::Boolean>().Value();
+      query_options.cursor_name = NULL;
+    }
+    else
+    {
+      Napi::TypeError::New(env, "Connection.query options: ." QUERY_OPTION_PROPERTY_NAME_CURSOR " must be a STRING or a BOOLEAN value.").ThrowAsJavaScriptException();
```
Comment (kadler) [inline src/odbc_connection.cpp:529]: Of course, making it a variable then you can't do this.

File: src/odbc_connection.cpp [hunk 3]
```diff
@@ -674,14 +471,156 @@ Napi::Value ODBCConnection::CreateStatement(const Napi::CallbackInfo& info) {
...
+  Napi::Env   env,
+  Napi::Value options_value
+)
+{
+  QueryOptions query_options;
+
+  if (options_value.IsNull())
+  {
+    query_options.use_cursor  = false;
+    query_options.cursor_name = NULL;
```
Comment (kadler) [inline src/odbc_connection.cpp:496]: These could be default initialized at object construction:

```C++
struct QueryOptions {
  bool use_cursor = false;
  SQLTCHAR* cursor_name = nullptr;
};
```

File: src/odbc_connection.cpp [hunk 4]
```diff
@@ -674,14 +471,156 @@ Napi::Value ODBCConnection::CreateStatement(const Napi::CallbackInfo& info) {
...
+)
+{
+  data->sqlReturnCode =
+  SQLSetStmtAttr
+  (  
+    data->hSTMT,  
+    SQL_ATTR_ROW_ARRAY_SIZE,  
+    (SQLPOINTER) fetch_size,  
+    0
+  );
```
Comment (kadler) [inline src/odbc_connection.cpp:591]: FYI, probably should check for warning 01S02 - Option value changed and call SQLGetStmtAttr to get the adjusted fetch size.

**Review by kadler — COMMENTED** [review]
First pass review. Would have been nice to have the QueryData -> StatementData change in a separate PR. Really adds a lot of noise to the changes.

File: src/odbc.h [hunk 5]
```diff
@@ -210,37 +231,39 @@ typedef struct QueryData {
...
           case SQL_C_SLONG:
-            delete[] (SQLUINTEGER *)this->boundRow[i];
+            delete[] (SQLUINTEGER *)this->bound_columns[i].buffer;
             break;
           case SQL_C_UBIGINT:
-            delete[] (SQLUBIGINT *)this->boundRow[i];
+            delete[] (SQLUBIGINT *)this->bound_columns[i].buffer;
             break;
         }
+
```
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc.h:259]: Vector is definitely sounding more and more like something that will have to be implemented for these buffers AND the `char_data` and `wchar_data` on the `ColumnData` struct.

Most of this comes from my affinity for C, and how I can squint and see `new[]`/`delete[]` as still feeling like C.

File: lib/Connection.js [hunk 6]
```diff
@@ -26,22 +27,45 @@ class Connection {
...
+    if (typeof options === 'undefined')
+    {
+      if (typeof parameters === 'object' && parameters !== null && !Array.isArray(parameters))
+      {
+        options = parameters;
         parameters = null;
-      } else if (typeof parameters === 'undefined') {
-        parameters = null;
+      } else {
+        options = null;
```
[the remark this answers is not in the document]
Comment (markdirish) [reply lib/Connection.js:68]: I agree its a huge headache. The biggest thing is maintaining API compatibility between this PR (and the eventual 2.4.0) and everything that came before that lacked the object. And parameters feel one-tier-of-importance above the rest of the query options that I can imagine, so it seems like if anything gets their own argument, it should be them.

All of these different combinations of arguments are covered in the tests, so I am fairly confident they are working as expected.

File: src/odbc_statement.cpp [hunk 7]
```diff
@@ -329,10 +329,18 @@ class ExecuteAsyncWorker : public ODBCAsyncWorker {
...
+    StatementData  *data;
 
     void Execute() {
+
+      // TODO: add debug stuff
+      set_fetch_size
+      (
+        data,
+        1
+      );
```
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc_statement.cpp:342]: I have been doing a _lot_ of experimentation with function call formatting. I'm kind of a fan of "the more vertical the better", with return value, function name, and arguments all on their own lines. In this case it seems like overkill, but there are a lot of places where it really helps me see whats going on at a glance.

File: src/odbc_connection.cpp (hunk 2 above)
Comment (markdirish) [reply src/odbc_connection.cpp:529]: Does this make any sense?

```
Napi::TypeError::New(env, std::string("Connection.query options: ." + std::string(QueryOptions::FETCH_SIZE_PROPERTY) + " must be a NUMBER value.")).ThrowAsJavaScriptException();
```

It compiles and works, but just wondering if its too ugly.

Comment (kadler) [reply src/odbc_connection.cpp:529]: You can append a `const char*` to a string:

```c++
Napi::TypeError::New(env, std::string("Connection.query options: .") + QueryOptions::FETCH_SIZE_PROPERTY + " must be a NUMBER value.")).ThrowAsJavaScriptException();
```

File: src/odbc_cursor.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+      Napi::Env env = Env();
+      Napi::HandleScope scope(env);
+
+      // TODO: Get the actual parameters
+      // Napi::Array empty = Napi::Array::New(env);
+      Napi::Array rows = process_data_for_napi(env, data, cursor->napiParametersReference.Value());
+
+      std::vector<napi_value> callbackArguments;
+      callbackArguments.push_back(env.Null());
+      callbackArguments.push_back(rows);
```
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc_cursor.cpp]: Really should be done for _ALL_ callbacks that come from the AsyncWorkers, but will create a different PR for that.

File: src/odbc.h [hunk 9]
```diff
@@ -77,6 +79,16 @@ typedef struct Column {
   SQLLEN        buffer_size; // size of the buffer bound
 } Column;
 
+typedef struct ColumnBuffer {
+  SQLPOINTER  buffer;
+  SQLLEN     *length_or_indicator_array;
+} ColumnBuffer;
```
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc.h:85]: My compiler won't compile without the `typedef` included in these structs. It errors:

```
../src/odbc.h:149:3: error: ‘ColumnBuffer’ does not name a type
   ColumnBuffer               *bound_columns = NULL;
```

File: src/odbc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc.h:258]: Will do this in a future PR, would clutter this one up some more.

File: src/odbc_statement.cpp [hunk 10]
```diff
@@ -347,8 +355,10 @@ class ExecuteAsyncWorker : public ODBCAsyncWorker {
       }
       DEBUG_PRINTF("[SQLHENV: %p][SQLHDBC: %p][SQLHSTMT: %p] ODBCStatement::ExecuteAsyncWorker::Execute(): SQLExecute succeeded\n", odbcConnection->hENV, odbcConnection->hDBC, data->hSTMT);
 
-      data->sqlReturnCode = this->odbcConnection->RetrieveResultSet(data);
-      if (!SQL_SUCCEEDED(data->sqlReturnCode)) {
+      data->sqlReturnCode = prepare_for_fetch(data);
```
[the remark this answers is not in the document]
Comment (markdirish) [reply src/odbc_statement.cpp:358]: Nah, just the pace of development. The top-level calls of an AsyncWorker::Execute() call need to call `SetError` if it errors to return the errors correctly.

**Pull request merged**